apps/rush-lib/src/logic/pnpm/PnpmProjectDependencyManifest.ts

```diff
@@ -0,0 +1,125 @@
+
+import * as path from 'path';
+import {
+  JsonFile,
+  InternalError,
+  FileSystem
+} from '@microsoft/node-core-library';
+
+import {
+  PnpmShrinkwrapFile,
+  IPnpmShrinkwrapDependencyYaml
+} from './PnpmShrinkwrapFile';
+import { RushConfigurationProject } from '../../api/RushConfigurationProject';
+import { RushConstants } from '../RushConstants';
+import { BasePackage } from '../base/BasePackage';
+
+export interface IPnpmProjectDependencyManifestOptions {
+  pnpmShrinkwrapFile: PnpmShrinkwrapFile;
+  project: RushConfigurationProject;
+}
+
+/**
+ * This class handles creating the project/.rush/temp/project-dependencies.json file
+ * which tracks the direct and indirect dependencies that a project consumes. This is used
+ * to better determine which projects should be rebuilt when dependencies are updated.
+ */
+export class PnpmProjectDependencyManifest {
+  private static _cache: Map<string, string> = new Map<string, string>();
+
+  private readonly _projectDependencyManifestFilename: string;
+  private readonly _pnpmShrinkwrapFile: PnpmShrinkwrapFile;
+  private readonly _project: RushConfigurationProject;
+
+  private _projectDependencyManifestFile: Map<string, string>;
+
+  /**
+   * Get the fully-qualified path to the project/.rush/temp/project-dependencies.json
+   * for the specified project.
+   */
+  public static getFilePathForProject(project: RushConfigurationProject): string {
+    return path.join(
+      project.projectRushTempFolder,
+      RushConstants.projectDependencyManifestFilename
+    );
+  }
+
+  public constructor(options: IPnpmProjectDependencyManifestOptions) {
+    this._pnpmShrinkwrapFile = options.pnpmShrinkwrapFile;
+    this._project = options.project;
+    this._projectDependencyManifestFilename = PnpmProjectDependencyManifest.getFilePathForProject(this._project);
+
+    this._projectDependencyManifestFile = new Map<string, string>();
+  }
+
+  public addDependency(pkg: BasePackage): void {
+    if (!pkg.version) {
+      throw new InternalError(`Version missing from dependency ${pkg.name}`);
+    }
+
+    this._addDependencyInternal(pkg.name, pkg.version);
+  }
+
+  /**
+   * Save the current state of the object to project/.rush/temp/project-dependencies.json
+   */
+  public save(): void {
+    const file: { [specifier: string]: string } = {};
+    const keys: string[] = Array.from(this._projectDependencyManifestFile.keys()).sort();
+    for (const key of keys) {
+      file[key] = this._projectDependencyManifestFile.get(key)!;
+    }
+
+    JsonFile.save(
+      file,
+      this._projectDependencyManifestFilename,
+      { ensureFolderExists: true }
+    );
+  }
+
+  /**
+   * If the project/.rush/temp/project-dependencies.json file exists, delete it. Otherwise, do nothing.
+   */
+  public deleteIfExists(): void {
+    FileSystem.deleteFile(this._projectDependencyManifestFilename, { throwIfNotExists: false });
+  }
+
+  private _addDependencyInternal(
```

Comment:
I see that the resulting file looks like this:

**my-project/.rush/temp/project-dependencies.json**
```ts
{
  "@types/node@8.10.54": "sha512-kaYyLYf6ICn6/isAyD4K1MyWWd5Q3JgH6bnMN089LUx88+s4W8GvK9Q6JMBVu5vsFFp7pMdSxdKmlBXwH/VFRg==",
  "colors@1.2.5": "sha512-erNRLao/Y3Fv54qUa0LBB+//Uf3YwMUmdJinN20yMXm9zdKKqH9wt7R9IIVZ+K7ShzfpLV/Zg8+VyrBJYB4lpg==",
  "gulp@4.0.2": "sha512-dvEs27SCZt2ibF29xYgmnwwCYZxdxhQ/+LFWlbAW8y7jt68L/65402Lz3+CKy0Ov4rOs+NERmDq7YlZaDqUIfA=="
}
```

What are these hashes computed from?  How does it handle peer dependencies and optional dependencies and such?  Is there a GitHub issue somewhere explaining the basic design?  (A doc comment for `PnpmProjectDependencyManifest` would be an okay place to discuss this as well.)

We want for people in the community to have an easy time understanding how this code works so they can diagnose problems and potentially contribute fixes in the future.

Also where's the PR description?  😊

Reply:
Those hashes come from PNPM's shrinkwrap file.

Reply:
I've added support for both. Optional dependencies was easier as that one was already resolved. The peerDependencies was just a semver, so what I do is go through the peerDependencies listed on a dependency, and check 3 things:
1. Does the package with the peerDependency contain a dependency that satisfies the peerDependency? If so, skip to the next peerDependency
2. If not, does the parent package contain a dependency that satisfies the peerDependency? If so, skip to the next peerDependency
3. If not, does the shrinkwrap have a top-level dependency that satisfies the peerDependency? If so, add this as a dependency to the project-dependencies.json. If not, blow up.

Reply:
@D4N14L I'm not sure this heuristic for peerDependencies is correct.

For example, `rush-lib` depends on `@pnpm/link-bins` which has a peer dependency on `@pnpm/logger`:

**rush-lib/package.json**
```js
{
  "name": "@microsoft/rush-lib",
  . . .
  "dependencies": {
    . . .
    "@pnpm/link-bins": "~1.0.1",
    "@pnpm/logger": "~1.0.1",
```

In the `pnpm-lock.json` file, it looks like this:

**common/temp/pnpm-lock.json**
```yaml
  'file:projects/rush-lib.tgz':
    dependencies:
      '@pnpm/link-bins': 1.0.3_@pnpm+logger@1.0.2
      '@pnpm/logger': 1.0.2
      . . .
    dev: false
    name: '@rush-temp/rush-lib'
    resolution:
      integrity: sha512-QsoGEBDl0OmtUvjAlxqQhxzSxcb81ue+rUkscwnG8wHRQKJAd3T0hj3EvwoBPxoxBWskiSU2hnflpzE8YEb3xw==
      tarball: 'file:projects/rush-lib.tgz'
    version: 0.0.0

  . . .

  /@pnpm/link-bins/1.0.3_@pnpm+logger@1.0.2:
    dependencies:
      '@pnpm/logger': 1.0.2
      '@pnpm/package-bins': 1.0.0
      '@pnpm/types': 1.8.0
      . . .
    peerDependencies:
      '@pnpm/logger': ^1.0.0
    resolution:
      integrity: sha512-thVgwrQ5rMcPYI6a0IPOt2pnlF1n5zX7BN4CrFeBp0/JCGsZAht/VOPv9bD3cZ+j0vDemEwE23BfhOWxmxq2yQ==
```

Thus the right thing to do is traverse into the entry for `/@pnpm/link-bins/1.0.3_@pnpm+logger@1.0.2` (and possibly any other installations of `@pnpm/link-bins` that may appear in the directed acyclic graph).  You shouldn't need to do any module resolution, since the **pnpm-lock.json** file already has graph completely worked out.

I think this is what Ian was saying above.

Reply:
We investigated this offline. @D4N14L pointed out that when a peer dependency is satisfied by a direct dependency (of the `common/temp/package.json` file), PNPM excludes it from the `dependencies:` list.  The reason is that the shrinkwrap file stores a plan for installing folders, and the installation plan relies on NodeJS module resolution to find the direct dependency.  So we're cool.